### PR TITLE
[Relay][Op] Dropout and batch_norm

### PR DESCRIPTION
**Pull request by slyubomirsky**
Dropout was fairly simple, but I had to make some decisions with `batch_norm`, so I would appreciate review. In particular, symbolic shapes do not work for the `batch_norm` relation I wrote because it needs to compare for equality (which forces them to be ints): Is that okay, or is there another way I can enforce those equalities?

### CODE REVIEW

File: src/relay/op/nn/nn.cc [hunk 1]
```diff
@@ -143,5 +143,162 @@ RELAY_REGISTER_UNARY_OP("relay.op.nn._make.", "relu")
...
+  CHECK(candidate != nullptr)
+    << name << " should be a vector but is not a tensor type,";
+  CHECK_EQ(dtype, candidate->dtype)
+    << name << " should be of the same data type as the original but it is not.";
+  CHECK_EQ(candidate->shape.size(), 1)
+    << name << " should be a vector but has a shape of "
+    << candidate->shape.size() << " dimensions instead of 1.";
+
+  auto length = as_const_int(candidate->shape[0]);
+  if (length == nullptr) return false;
```
Comment (slyubomirsky) [inline src/relay/op/nn/nn.cc]: Would using the TypeReporter's `AssertEQ` method work here for handling the case of symbolic dimensions? The method's implementation seems like a placeholder for now. I think in principle we'd want to have a constraint that the dimensions match and it shouldn't require coercing them into constants immediately

**Comment by slyubomirsky** [comment]
Another question about `batch_norm`: The NNVM docs said that the default axis argument was 1. Was that 0-based or 1-based?

File: src/relay/op/nn/nn.cc [hunk 2]
```diff
@@ -143,5 +143,162 @@ RELAY_REGISTER_UNARY_OP("relay.op.nn._make.", "relu")
...
+.describe(R"code(Applies the dropout operation to the input array.
+
+During training, each element of the input is set to zero with probability ``p``.
+The whole array is rescaled by ``1/(1-p)`` to keep the expected sum of the input unchanged.
+
+)code" TVM_ADD_FILELINE)
+.set_num_inputs(1)
+.add_argument("data", "Tensor", "Input to which dropout will be applied.")
+.set_support_level(1)
+.add_type_rel("Identity", IdentityRel);
```
Comment (siju-samuel) [inline src/relay/op/nn/nn.cc]: Dropout has 2 outputs,  I think you have to add as TupleTypeNode

File: tests/python/relay/test_op_level1.py [hunk 3]
```diff
@@ -169,6 +169,91 @@ def test_concatenate_infer_type():
...
+    ib = relay.ir_builder.IRBuilder()
+    n, t, d = tvm.var("n"), tvm.var("t"), tvm.var("d")
+    x = ib.param("x", relay.ty.TensorType((n, t, d), "float32"))
+    with ib.function(x) as func:
+        ib.ret(relay.nn.dropout(x, rate=0.75))
+    ib.ret(func)
+
+    func = relay.ir_pass.infer_type(ib.env, func.to_func())
+    ftype = func.checked_type
+    assert ftype.ret_type == relay.ty.TensorType((n, t, d), "float32")
```
Comment (siju-samuel) [inline tests/python/relay/test_op_level1.py]: add validation for both the dropout output

File: src/relay/op/nn/nn.cc (hunk 2 above)
Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: What is the second output? The NNVM doc for dropout only makes reference to a single result tensor, but perhaps I'm missing something.

**Comment by slyubomirsky** [comment]
I'll remove the submodule then, pardon that

File: src/relay/op/nn/nn.cc (hunk 2 above)
Comment (siju-samuel) [reply src/relay/op/nn/nn.cc]: https://github.com/dmlc/tvm/blob/4e309e67a6c23411eec941000e469bd4e17cef92/nnvm/src/top/nn/nn.cc#L167

https://github.com/dmlc/tvm/blob/4e309e67a6c23411eec941000e469bd4e17cef92/nnvm/src/top/nn/nn.cc#L175

Here I can see 2 outputs, output and mask.

Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: Ah, you're right. The registry also says that only one of the outputs is visible, so presumably the "mask" output isn't exposed. I see no reference to the mask output anywhere else, so what characteristics should it have?

Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: Ah well I can infer based on `less` and `greater` that the mask tensor should have the same shape. I'll  put it in and see if other reviewers think this output should be exposed.

File: src/relay/op/nn/nn.cc [hunk 4]
```diff
@@ -146,6 +146,23 @@ RELAY_REGISTER_UNARY_OP("relay.op.nn._make.", "relu")
...
+  CHECK_EQ(types.size(), 2);
+  const auto* data = types[0].as<TensorTypeNode>();
+  if (data == nullptr) return false;
+
+  // dropout returns the original tensor with dropout applied
+  // and a mask tensor (1.0 where element not dropped, 0.0 where dropped)
+  std::vector<Type> fields;
+  fields.push_back(TensorTypeNode::make(data->shape, data->dtype));
+  fields.push_back(TensorTypeNode::make(data->shape, data->dtype));
+  reporter->Assign(types[1], TupleTypeNode::make(Array<Type>(fields)));
```
Comment (MarisaKirisame) [inline src/relay/op/nn/nn.cc]: You can try using TupleTypeNode::make(Array<Type>({ty, ty}));

Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: Duly noted. I've had some type errors with similar constructions, so I'll give it another shot.

**Comment by tqchen** [comment]
default axis=1 is 0 based indexing, and that is because usually batch_norm normalizes the channel dimension

**Comment by slyubomirsky** [comment]
Thanks for the explanation; I've changed the default

File: include/tvm/relay/attrs/transform.h [hunk 5]
```diff
@@ -1,87 +1,87 @@
...
-  }
-};  // struct FullAttrs
-
-}  // namespace relay
-}  // namespace tvm
-#endif  // TVM_RELAY_ATTRS_TRANSFORM_H_
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file tvm/relay/attrs/transform.h
+ * \brief Transform operators.
```
[the remark this answers is not in the document]
Comment (slyubomirsky) [reply include/tvm/relay/attrs/transform.h]: Oh dear. I'll look into that.

File: src/relay/op/nn/nn.cc [hunk 6]
```diff
@@ -217,5 +217,177 @@ Normalizes along dimension axis using an L2 norm
...
+)code" TVM_ADD_FILELINE)
+.set_num_inputs(1)
+.add_argument("data", "Tensor", "Input to which dropout will be applied.")
+.set_support_level(1)
+.add_type_rel("Dropout", DropoutRel);
+
+// batch_norm
+TVM_REGISTER_NODE_TYPE(BatchNormAttrs);
+
+bool CheckVectorLength(int64_t dim, const DataType& dtype, Type vector, std::string name) {
```
Comment (tqchen) [inline src/relay/op/nn/nn.cc]: use const char* for name so the constant string is constructed lazily

File: src/relay/op/nn/nn.cc [hunk 7]
```diff
@@ -217,5 +217,177 @@ Normalizes along dimension axis using an L2 norm
...
+  const auto* candidate = vector.as<TensorTypeNode>();
+  CHECK(candidate != nullptr)
+    << name << " should be a vector but is not a tensor type,";
+  CHECK_EQ(dtype, candidate->dtype)
+    << name << " should be of the same data type as the original but it is not.";
+  CHECK_EQ(candidate->shape.size(), 1)
+    << name << " should be a vector but has a shape of "
+    << candidate->shape.size() << " dimensions instead of 1.";
+
+  auto length = as_const_int(candidate->shape[0]);
```
Comment (tqchen) [inline src/relay/op/nn/nn.cc]: const auto* to give enough hint of the type here or even better const int64_t* as it is not too long to type

File: src/relay/op/nn/nn.cc (hunk 6 above)
Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: Thanks for the advice, I'll change it here and remember for the future.

File: src/relay/op/nn/nn.cc (hunk 7 above)
Comment (slyubomirsky) [reply src/relay/op/nn/nn.cc]: Duly noted

**Comment by tqchen** [comment]
Some final comments. Also it would also be really nice if we return a wrapper class in python to support the following sugar

```python
x, mask = relay.nn.dropout(input)
```

In order to support this, we could do the following trick
```python
class TupleExpr(tvm.node._ffi.NodeGeneric):
    def __init__(self, tuple_value, size):
        self.tuple_value = tuple_value
        self.size = size
   # pass the tuple value if itself  get passed as argument to a FFI function
   def asnode(self):
        return self.tuple_value
    # implement __getitem__, __len__, so it can behave like an array
```
And in the python function we wrap the return value by TupleExpr(result_of_dropout, 2)

**Comment by slyubomirsky** [comment]
Good idea, I'll add the tuple wrapper and use it for these functions

File: tests/python/relay/test_op_level1.py [hunk 8]
```diff
@@ -193,6 +193,93 @@ def test_l2_normalize():
...
+    func = relay.ir_pass.infer_type(ib.env, func.to_func())
+    ftype = func.checked_type
+    assert ftype.ret_type == relay.ty.TupleType([input_ty, input_ty])
+
+    ib = relay.ir_builder.IRBuilder()
+    n, t, d = tvm.var("n"), tvm.var("t"), tvm.var("d")
+    input_ty = relay.ty.TensorType((n, t, d), "float32")
+    x = ib.param("x", input_ty)
+    with ib.function(x) as func:
+        ib.ret(relay.nn.dropout(x, rate=0.75).asnode())
```
[the remark this answers is not in the document]
Comment (tqchen) [reply tests/python/relay/test_op_level1.py]: The ret function an be updated to be aware of NodeGeneric and call asnode automatically

**Comment by tqchen** [comment]
Thanks @slyubomirsky @siju-samuel , this is now merged
